src/gateway/GatewayV2.sol

```diff
@@ -0,0 +1,343 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+pragma solidity 0.8.21;
+
+import {Auth} from "src/Auth.sol";
+import {ArrayLib} from "src/libraries/ArrayLib.sol";
+import {BytesLib} from "src/libraries/BytesLib.sol";
+import {MessagesLib} from "src/libraries/MessagesLib.sol";
+import {SafeTransferLib} from "src/libraries/SafeTransferLib.sol";
+import {IGatewayV2} from "src/interfaces/gateway/IGatewayV2.sol";
+
+interface ManagerLike {
+    function handle(bytes memory message) external;
+}
+
+interface GasServiceLike {
+    function estimate(bytes calldata payload) external view returns (uint256);
+    function shouldRefuel(address source, bytes calldata payload) external view returns (bool);
+}
+
+interface RouterLike {
+    function send(bytes memory message) external;
+    function pay(bytes calldata payload, address refund) external payable;
+    function estimate(bytes calldata payload, uint256 destChainCost) external view returns (uint256);
+}
+
+interface RootLike {
+    function paused() external returns (bool);
+    function scheduleRely(address target) external;
+    function cancelRely(address target) external;
+    function recoverTokens(address target, address token, address to, uint256 amount) external;
+    function endorsed(address user) external view returns (bool);
+}
+
+/// @title  Gateway
+/// @notice Routing contract that forwards to multiple routers (1 full message, n-1 proofs)
+///         and validates multiple routers have confirmed a message.
+///
+///         Supports processing multiple duplicate messages in parallel by
+///         storing counts of messages and proofs that have been received.
+contract GatewayV2 is Auth, IGatewayV2 {
+    using ArrayLib for uint16[8];
+    using BytesLib for bytes;
+
+    uint8 public constant MAX_ROUTER_COUNT = 8;
+    uint8 public constant PRIMARY_ROUTER_ID = 1;
+    uint256 public constant RECOVERY_CHALLENGE_PERIOD = 7 days;
+    RootLike public immutable root;
+
+    address public investmentManager;
+    address public poolManager;
+    GasServiceLike public gasService;
+
+    address[] public routers;
+    mapping(address router => Router) public activeRouters;
+    mapping(bytes32 messageHash => Message) public messages;
+    mapping(bytes32 messageHash => Recovery) public recoveries;
+    mapping(uint8 messageId => address manager) messageHandlers;
+
+    uint256 allowance;
```

Comment:
I'm not a huge fan of the name allowance since it overlaps with ERC20 allowance/approvals. What about `quota`? Or even just `gasAllowance`

Reply:
I like `quota`. I also had some doubts about the allowance